tao_compiler/mlir/disc/transforms/shape_utils.cc

```diff
@@ -0,0 +1,1743 @@
+
+namespace mlir {
+namespace disc_ral {
+
+// Supports using EquivalenceClasses for Value
+bool operator<(const ValueWrapper& lhs, const ValueWrapper& rhs) {
+  auto lhs_value = lhs.getValue().getAsOpaquePointer();
+  auto rhs_value = rhs.getValue().getAsOpaquePointer();
+  return lhs_value < rhs_value;
+}
+
+// Merge two symbolDim if they are compatible.
+LogicalResult SymbolDim::Merge(SymbolDim* other) {
+  if (!isDynamic() && !other->isDynamic() &&
+      getDimSize() != other->getDimSize())
+    return failure();
+  if (isDynamic()) dimSize_ = other->getDimSize();
+  return success();
+}
+
+Type ShapeAnalysis::getRefinedType(Value value) {
+  SymbolShape* symbolShape = getShape(value);
+
+  // not-shaped type
+  if (!symbolShape) return value.getType();
+
+  auto ty = value.getType().cast<RankedTensorType>();
+  auto dims = symbolShape->getDims();
+  return RankedTensorType::get(dims, ty.getElementType());
+}
+
+LogicalResult ShapeAnalysis::run() {
+  // Make sure only run once.
+  if (initialized) {
+    op_->emitError() << "re-initialized shape analysis";
+    return failure();
+  }
+  initialized = true;
+
+  return buildShapeMap();
+}
+
+LogicalResult ShapeAnalysis::buildShapeMap() {
+  // Do not merge the region loops. We want to compute shape-map, dim-value-map
+  // and decompose one-by-one.
+  for (auto& region : op_->getRegions()) {
+    if (failed(buildRegionShapeMap(&region))) {
+      return failure();
+    }
+  }
+  for (auto& region : op_->getRegions()) {
+    if (failed(buildRegionDimValueMap(&region))) {
+      return failure();
+    }
+  }
+  for (auto& region : op_->getRegions()) {
+    if (failed(buildRegionMulDecompose(&region))) {
+      return failure();
+    }
+  }
+  postProcessingDimValues();
+
+  return success();
+}
+
+void ShapeAnalysis::postProcessingDimValues() {
+  // Update dimValue roots.
+  for (auto sym2val : dimSymbol2DimValue_) {
+    dimSymbol2DimValue_[sym2val.first] = getRootDimValue(sym2val.second);
+  }
+  std::unordered_map<DimValue, SmallVector<std::vector<DimValue>>, DimValueHash>
+      dimValMulDecomp;
+  for (auto delinear : dimValueMulDecompose_) {
+    auto key = getRootDimValue(delinear.first);
+    std::unordered_set<std::vector<DimValue>, DimValueContainerHash>
+        decompDimVals;
+    for (auto val_set : delinear.second) {
+      std::vector<DimValue> vals;
+      for (auto val : val_set) {
+        vals.push_back(getRootDimValue(val));
+      }
+      std::sort(vals.begin(), vals.end());
+      decompDimVals.insert(vals);
+    }
+    dimValMulDecomp[key] = SmallVector<std::vector<DimValue>>(
+        decompDimVals.begin(), decompDimVals.end());
+  }
+  dimValueMulDecompose_ = std::move(dimValMulDecomp);
+  // Analyze multi-level decompose:
+  //   a -> b, c
+  //   b -> d, e
+  //      a -> d, e, c
+  // Note if `c` is one, we can get: a -> b, c, c, c, ...
+  // We add a constraint that every element can be decomposed to as many as 4
+  // elements.
```

Comment:
I kind of feel the overall process dealing with reshape is not so straightforward and quite hard to understand.
Maybe we should add some comments, providing an example with reshape and interpreting what analysis will be done in each phase.

BTW, just for some discussion, will it be even more simpler if adopting symbolic calculation?